### PR TITLE
35 class parser

### PR DESCRIPTION
**Pull request by pontuslaestadius**
Deserialization of json to java object (parsing)
Send to the backend
Extending diagramCheck
Fix ParralelDiagram

### CODE REVIEW

File: Application/src/model/Parser.java [hunk 1]
```diff
@@ -35,60 +41,85 @@
...
+    private String diagram;
+    private String parallel;
 
-    private String tempParallelDiagram = "";
-    private String parallelDiagramString = "";
+    /**
+     * Parses a class diagram.
+     * @param inputJSON to parse
+     */
+    void parseClassDiagram(String inputJSON) {
```
Comment (ghost) [inline Application/src/model/Parser.java]: should this be public?

File: Application/src/model/Parser.java [hunk 2]
```diff
@@ -99,66 +130,65 @@ public void parseSequenceDiagram(String inputJSON) {
...
 
-            // get the elements of the First Diagram
-            for (Messages diagramElement : parsedSequenceDiagram.getDiagram().getContent().get(0).getMessages()) {
-                tempFirstDiagram += "{" + diagramElement.getFrom() + "," + diagramElement.getTo() + ",[" + diagramElement.getMessage().get(0) + ", " +
-                        diagramElement.getMessage().get(1) + ", " + diagramElement.getMessage().get(2) + "]" + "}" + ",";
-                //removing extra comma
-                firstDiagramString = (tempFirstDiagram.substring(0, tempFirstDiagram.length() - 1));
-            }
-
+            diagram = parseMessages(parsedSequenceDiagram.getDiagram().getContent().get(0).getMessages());
```
Comment (ghost) [inline Application/src/model/Parser.java]: add a comment here please

File: Application/src/model/Parser.java [hunk 3]
```diff
@@ -99,66 +130,65 @@ public void parseSequenceDiagram(String inputJSON) {
...
+            diagram = wrap(diagram);
 
-            // get the elements of the Parallel Diagram
-            for (Messages parallelDiagramElement : parsedSequenceDiagram.getDiagram().getContent().get(1).getMessages()) {
-                tempParallelDiagram += "{" + parallelDiagramElement.getFrom() + "," + parallelDiagramElement.getTo() + ",[" + parallelDiagramElement.getMessage().get(0) + "]" + "}" + ",";
-                //removing extra comma
-                parallelDiagramString = (tempParallelDiagram.substring(0, tempParallelDiagram.length() - 1));
+            // If there is a parallel diagram.
+            if (parsedSequenceDiagram.getDiagram().getContent().size() > 1) {
+                parallel = parseMessages(parsedSequenceDiagram.getDiagram().getContent().get(1).getMessages());
```
Comment (ghost) [inline Application/src/model/Parser.java:146]: i think this is fine, but @timjon  expressed his concern about the possibility of there being more than one parallel diagram. we should bring this up in the next physical meeting

File: Application/src/model/classDiagram/Relationships.java [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+public class Relationships {
+    public String type;
+    public String superclass;
+    public String subclass;
+
+    /**
+     * @return a formatted string.
+     */
+    public String format() {
+        return "[" + type + "," + superclass + "," + subclass + "]";
```
Comment (ghost) [inline Application/src/model/classDiagram/Relationships.java:17]: good idea!

File: Application/src/model/DiagramCheck.java [hunk 5]
```diff
@@ -1,57 +1,54 @@
...
+ * collaborator: Pontus Laestadius
  * @since 2017-10-10
  */
 
-public class DiagramCheck {
-	// Supported diagram types.
+class DiagramCheck {
+
+    // Supported diagram types.
+    private final static String[] allowedDiagrams = {"sequence_diagram", "class_diagram", "deployment_diagram"};
```
Comment (ghost) [inline Application/src/model/DiagramCheck.java:17]: you don't have to change this, but maybe add deployment_diagram later when we work on it? :)

File: Application/src/model/Parser.java [hunk 6]
```diff
@@ -35,60 +41,85 @@
...
+     * Parses a class diagram.
+     * @param inputJSON to parse
+     */
+    void parseClassDiagram(String inputJSON) {
+
+        try {
 
+            // Parse
+            Gson gson = new Gson();
+            ClassDiagram cd = gson.fromJson(inputJSON, ClassDiagram.class);
```
Comment (ghost) [inline Application/src/model/Parser.java]: needs a consistent instance variable name, e.g.  parsedClassDiagram

File: Application/src/model/Parser.java [hunk 7]
```diff
@@ -35,60 +41,85 @@
 
 public class Parser {
 
-    /**
+    /*
      * defining two different counter variables that will increment each time the diagram calling method
      * that they are is called in order to differentiate between the imported diagrams
      * These two counters are static in order that they change each time and new diagram is imported from the main method
```
Comment (ghost) [inline Application/src/model/Parser.java]: since you don't use the old counters anymore, don't forget to delete this comment

File: Application/src/model/Parser.java [hunk 8]
```diff
@@ -35,60 +41,85 @@
...
 
     /**
      * Parses the imported JSON files if they contain a sequence diagram adheres with the predetermined JSON format
      *
      * @param inputJSON the imported collection of Strings that make up the JSON files
      */
-
     public void parseSequenceDiagram(String inputJSON) {
+        String tempProcesses = "";
+        String tempClassNames = "";
```
Comment (ghost) [inline Application/src/model/Parser.java:116]: should these have a private declaration?

File: Application/src/model/Parser.java [hunk 9]
```diff
@@ -99,66 +130,65 @@ public void parseSequenceDiagram(String inputJSON) {
...
+            diagram = parseMessages(parsedSequenceDiagram.getDiagram().getContent().get(0).getMessages());
+            diagram = wrap(diagram);
 
-            // get the elements of the Parallel Diagram
-            for (Messages parallelDiagramElement : parsedSequenceDiagram.getDiagram().getContent().get(1).getMessages()) {
-                tempParallelDiagram += "{" + parallelDiagramElement.getFrom() + "," + parallelDiagramElement.getTo() + ",[" + parallelDiagramElement.getMessage().get(0) + "]" + "}" + ",";
-                //removing extra comma
-                parallelDiagramString = (tempParallelDiagram.substring(0, tempParallelDiagram.length() - 1));
+            // If there is a parallel diagram.
+            if (parsedSequenceDiagram.getDiagram().getContent().size() > 1) {
```
Comment (ghost) [inline Application/src/model/Parser.java:145]: i have a little suggestions over here
you are checking there is more than one diagram, so you are assuming the the second diagram is always a parallel diagram, try using the provided node to be more accurate
if (parsedSequenceDiagram.getDiagram().getNode() == "par")
take another look at the provided sequence diagram and you will find the par node

File: Application/src/controller/network/Net.java [hunk 10]
```diff
@@ -102,8 +102,15 @@ private void inner_loop() {
                     // Run the decode on the Platform instance.
                     Platform.runLater(() -> {
 
+                        // Pull a message from the received messages queue.
+                        String message = received.poll();
+
+                        // Due to threading, messages can be pulled and there can not be any.
```
Comment (kosara-g) [inline Application/src/controller/network/Net.java:108]: this comment is a bit ambiguous

File: Application/src/model/DiagramCheck.java (hunk 5 above)
Comment (kosara-g) [reply Application/src/model/DiagramCheck.java:17]: no, I think that's fine, we don't know what else we will have time for this sprint

**Review by ghost — APPROVED** [review]

**Review by kosara-g — APPROVED** [review]

File: Application/src/controller/network/Net.java [hunk 11]
```diff
@@ -102,8 +102,15 @@ private void inner_loop() {
                     // Run the decode on the Platform instance.
                     Platform.runLater(() -> {
 
+                        // Pull a message from the received messages queue.
```
Comment (timjon) [inline Application/src/controller/network/Net.java:105]: Why do you have changes in this file?

File: Application/src/model/DiagramCheck.java [hunk 12]
```diff
@@ -1,57 +1,54 @@
...
- * @version 0.1
+ * @version 1.0
  * @author Tim Jonasson & Kosara Golemshinska
+ * collaborator: Pontus Laestadius
  * @since 2017-10-10
  */
 
-public class DiagramCheck {
-	// Supported diagram types.
+class DiagramCheck {
```
Comment (timjon) [inline Application/src/model/DiagramCheck.java:14]: Why is this changed?

File: Application/src/model/Menu.java [hunk 13]
```diff
@@ -163,29 +163,39 @@ public static void pause() {
...
+                        Net.push(parse.getDiagram());
+                        Net.push(parse.getParallelSequenceDiagram());
+
+                        // Enable all media buttons.
+                        setMenuState(true);
+                        button_previous.setDisable(true);
+                        break;
+                    case "class_diagram":
+                        parse.parseClassDiagram(file);
+                        System.out.println(parse.getDiagram()); // TODO print until backend is implemented.
```
Comment (timjon) [inline Application/src/model/Menu.java:199]: So the user story is not done?

**Review by timjon — CHANGES_REQUESTED** [review]

File: Application/src/model/Menu.java (hunk 13 above)
Comment (ghost) [reply Application/src/model/Menu.java:199]: @timjon the story is done, this line just shows what parsing returns to facilitate integration, i don't think #33 should be blocked

File: Application/src/controller/network/Net.java (hunk 11 above)
Comment (pontuslaestadius) [reply Application/src/controller/network/Net.java:105]: I found a bug which was occurring when you try to open multiple diagrams at once i.e. a sequence and it's parallel at the same time. It would run out of items in the queue while still pulling.

Comment (pontuslaestadius) [reply Application/src/controller/network/Net.java:105]: So it's very related to my feature.

**Comment by timjon** [comment]
Breaks if there are any errors in the syntax

**Review by timjon — APPROVED** [review]
